src/Functions/punycode.cpp

```diff
@@ -0,0 +1,199 @@
+#include "config.h"
+
+#if USE_IDNA
+
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionStringToString.h>
+
+#ifdef __clang__
+#    pragma clang diagnostic push
+#    pragma clang diagnostic ignored "-Wnewline-eof"
+#endif
+#    include <ada/idna/punycode.h>
+#    include <ada/idna/unicode_transcoding.h>
+#ifdef __clang__
+#    pragma clang diagnostic pop
+#endif
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+    extern const int ILLEGAL_COLUMN;
+}
+
+struct PunycodeEncodeImpl
+{
+    static void vector(
+        const ColumnString::Chars & data,
+        const ColumnString::Offsets & offsets,
+        ColumnString::Chars & res_data,
+        ColumnString::Offsets & res_offsets)
+    {
+        const size_t rows = offsets.size();
+        res_data.resize(rows * 64); /// just a guess
+        res_offsets.resize(rows);
```

Comment:
`res_data` and `res_offset` both have `reserve()` functions, i.e. they can preallocate memory. So we don't really have to do that here manually and we can simplify the code.

Reply:
All manual memory management is now gone. `prev_res_offset` is also gone. But I don't see a prettier alternative to `prev_offset` (--> kept it).